ccl/backupccl: TestRestoreOldVersions failed

ccl/backupccl.TestRestoreOldVersions [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=3498673&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=3498673&tab=artifacts#/) on master @ [ecaa4920192549af1418923648769ede3de1c9c8](https://github.com/cockroachdb/cockroach/commits/ecaa4920192549af1418923648769ede3de1c9c8):

Fatal error:

```
panic: mutex is not write locked [recovered]
	panic: mutex is not write locked [recovered]
	panic: mutex is not write locked

```
Stack: 

```
goroutine 725815 [running]:
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).Recover(0xc002f51440, 0xa977a30, 0xc00b0fcfc0)
	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:231 +0x15d
panic(0x6733640, 0xa8830a0)
	/usr/local/go/src/runtime/panic.go:971 +0x499
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Store).Send.func1(0xc032b868e0, 0xc032b869c0, 0xc0059c4900, 0xc032b869b8)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/store_send.go:95 +0x745
panic(0x6733640, 0xa8830a0)
	/usr/local/go/src/runtime/panic.go:971 +0x499
github.com/cockroachdb/cockroach/pkg/util/syncutil.(*Mutex).AssertHeld(...)
	/go/src/github.com/cockroachdb/cockroach/pkg/util/syncutil/mutex_sync_race.go:49
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).cleanupFailedProposalLocked(0xc0066a5500, 0xc00666f8c0)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/replica.go:674 +0x1a5
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*replicaProposer).rejectProposalWithRedirectLocked(0xc0066a5500, 0xa977a30, 0xc00b0fd0e0, 0xc00666f8c0, 0x6)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/replica_proposal_buf.go:1056 +0x325
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*propBuf).FlushLockedWithRaftGroup(0xc0066a5b18, 0xa977a30, 0xc00b0fd0e0, 0xa953ce0, 0xc0097745d0, 0x0, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/replica_proposal_buf.go:453 +0x10fe
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*propBuf).flushLocked.func1(0xa953ce0, 0xc0097745d0, 0xc009948d80, 0xc007a53580)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/replica_proposal_buf.go:340 +0x6c
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*replicaProposer).withGroupLocked.func1(0xc0097745d0, 0x18, 0xc00dc5bc20, 0x1)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/replica_proposal_buf.go:996 +0x74
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).withRaftGroupLocked.func1(...)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/replica_raft.go:1504
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).withRaftGroupLocked(0xc0066a5500, 0x643b101, 0xc032b838d8, 0xc00c43ce80, 0xc032b838f0)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/replica_raft.go:1505 +0xea
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*replicaProposer).withGroupLocked(0xc0066a5500, 0xc00c43ce80, 0x2ccc8dd, 0xc0066a5940)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/replica_proposal_buf.go:992 +0x78
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*propBuf).flushLocked(0xc0066a5b18, 0xa977a30, 0xc00b0fd0e0, 0x0, 0x3)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/replica_proposal_buf.go:339 +0xf5
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*propBuf).flushRLocked(0xc0066a5b18, 0xa977a30, 0xc00b0fd0e0, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/replica_proposal_buf.go:335 +0x205
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*propBuf).allocateIndex(0xc0066a5b18, 0xa977a30, 0xc00b0fd0e0, 0x2997a00, 0xc014087218, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/replica_proposal_buf.go:297 +0x15b
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*propBuf).Insert(0xc0066a5b18, 0xa977a30, 0xc00b0fd0e0, 0xc000f99440, 0x1, 0x0, 0x0, 0x0, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/replica_proposal_buf.go:231 +0x1b1
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).propose(0xc0066a5500, 0xa977a30, 0xc00b0fd0e0, 0xc000f99440, 0x1, 0x0, 0x0, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/replica_raft.go:363 +0x6a5
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).evalAndPropose(0xc0066a5500, 0xa977a30, 0xc00b0fd0e0, 0xc02407b5e0, 0xc00b202960, 0x16a849faeb09cdee, 0x0, 0xc029548600, 0x900000009, 0x4, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/replica_raft.go:210 +0x905
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).executeWriteBatch(0xc0066a5500, 0xa977a30, 0xc00b0fd0e0, 0xc02407b5e0, 0xc00b202960, 0x0, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/replica_write.go:143 +0x72e
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).executeBatchWithConcurrencyRetries(0xc0066a5500, 0xa977a30, 0xc00b0fd0e0, 0xc02407b5e0, 0x760aff0, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/replica_send.go:429 +0x487
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).sendWithRangeID(0xc0066a5500, 0xa977a30, 0xc00b0fd0b0, 0x2, 0xc02407b5e0, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/replica_send.go:159 +0x68e
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).Send(...)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/replica_send.go:95
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Store).Send(0xc0059c4900, 0xa977a30, 0xc00b0fd080, 0x16a84a06d8ab6539, 0x0, 0x0, 0x900000009, 0x4, 0x0, 0x2, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/store_send.go:190 +0x5d8
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Stores).Send(0xc008b85080, 0xa977a30, 0xc00b0fd080, 0x0, 0x0, 0x0, 0x900000009, 0x4, 0x0, 0x2, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/stores.go:191 +0x1a5
github.com/cockroachdb/cockroach/pkg/server.(*Node).batchInternal.func1(0xa977a30, 0xc00b0fcfc0, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/server/node.go:930 +0x258
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunTaskWithErr(0xc002f51440, 0xa977a30, 0xc00b0fcfc0, 0x6f95509, 0x10, 0xc0050d1078, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:324 +0x12a
github.com/cockroachdb/cockroach/pkg/server.(*Node).batchInternal(0xc008e1ee00, 0xa977a30, 0xc00b0fcfc0, 0x1, 0xc02407b340, 0x0, 0x1, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/server/node.go:918 +0x285
github.com/cockroachdb/cockroach/pkg/server.(*Node).Batch(0xc008e1ee00, 0xa977a30, 0xc00b0fccc0, 0xc02407b340, 0xc008e1ee00, 0xc00d273f48, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/server/node.go:1025 +0x127
github.com/cockroachdb/cockroach/pkg/roachpb._Internal_Batch_Handler.func1(0xa977a30, 0xc00b0fccc0, 0x6f445c0, 0xc02407b340, 0x0, 0x0, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/roachpb/api.pb.go:8719 +0xa2
github.com/cockroachdb/cockroach/pkg/util/tracing.ServerInterceptor.func1(0xa977a30, 0xc00b0fccc0, 0x6f445c0, 0xc02407b340, 0xc0223303c0, 0xc024185b18, 0x0, 0x0, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/util/tracing/grpc_interceptor.go:126 +0x67d
google.golang.org/grpc.getChainUnaryHandler.func1(0xa977a30, 0xc00b0fccc0, 0x6f445c0, 0xc02407b340, 0x6c689c0, 0x1, 0xc012b717c0, 0xc0050d1688)
	/go/src/github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc/server.go:1134 +0x150
github.com/cockroachdb/cockroach/pkg/rpc.NewServer.func1(0xa977a30, 0xc00b0fccc0, 0x6f445c0, 0xc02407b340, 0xc0223303c0, 0xc012b717c0, 0xc012b717c0, 0x0, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/rpc/context.go:178 +0x125
google.golang.org/grpc.getChainUnaryHandler.func1(0xa977a30, 0xc00b0fccc0, 0x6f445c0, 0xc02407b340, 0x0, 0x0, 0xfacdbb, 0x40)
	/go/src/github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc/server.go:1134 +0x150
github.com/cockroachdb/cockroach/pkg/rpc.kvAuth.unaryInterceptor(0x1, 0xa977a30, 0xc00b0fccc0, 0x6f445c0, 0xc02407b340, 0xc0223303c0, 0xc012b71780, 0xc0050d1848, 0xf8c72c, 0xc012b717b0, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/rpc/auth.go:71 +0x139
google.golang.org/grpc.chainUnaryServerInterceptors.func1(0xa977a30, 0xc00b0fccc0, 0x6f445c0, 0xc02407b340, 0xc0223303c0, 0xc024185b18, 0xc02a8eb948, 0x4393d8, 0x30, 0x6c85480)
	/go/src/github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc/server.go:1120 +0x113
github.com/cockroachdb/cockroach/pkg/roachpb._Internal_Batch_Handler(0x6ef8260, 0xc008e1ee00, 0xa977a30, 0xc00b0fccc0, 0xc0125c3500, 0xc001255440, 0xa977a30, 0xc00b0fccc0, 0xc0109fc800, 0xba)
	/go/src/github.com/cockroachdb/cockroach/pkg/roachpb/api.pb.go:8721 +0x1f2
google.golang.org/grpc.(*Server).processUnaryRPC(0xc008a35500, 0xa9ddbb8, 0xc001c46f00, 0xc00c088ea0, 0xc007e17110, 0xcce3aa0, 0x0, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc/server.go:1293 +0xa66
google.golang.org/grpc.(*Server).handleStream(0xc008a35500, 0xa9ddbb8, 0xc001c46f00, 0xc00c088ea0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc/server.go:1618 +0x142c
google.golang.org/grpc.(*Server).serveStreams.func1.2(0xc0047abcc0, 0xc008a35500, 0xa9ddbb8, 0xc001c46f00, 0xc00c088ea0)
	/go/src/github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc/server.go:941 +0xe7
created by google.golang.org/grpc.(*Server).serveStreams.func1
	/go/src/github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc/server.go:939 +0x355
```
<details><summary>Log preceding fatal error</summary>
<p>

```
=== RUN   TestRestoreOldVersions/multi-region-restore/v21.1
```
</p>
</details>
<details><summary>Reproduce</summary>
<p>
To reproduce, try:

```bash
make stressrace TESTS=TestRestoreOldVersions PKG=./pkg/ccl/backupccl TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```
Parameters in this failure:

- GOFLAGS=-race -parallel=4
</p>
</details>
/cc @cockroachdb/bulk-io ajwerner
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestRestoreOldVersions.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

--- a/pkg/kv/kvserver/replica.go
+++ b/pkg/kv/kvserver/replica.go
@@ -408,9 +408,6 @@ type Replica struct {
 		// to the *RaftCommand contained in its associated *ProposalData. This
 		// is because the *RaftCommand can be mutated during reproposals by
 		// Replica.tryReproposeWithNewLeaseIndex.
-		//
-		// TODO(ajwerner): move the proposal map and ProposalData entirely under
-		// the raftMu.
 		proposals map[kvserverbase.CmdIDKey]*ProposalData
 		// Indicates that the replica is in the process of applying log entries.
 		// Updated to true in handleRaftReady before entries are removed from
@@ -669,9 +666,8 @@ func (r *Replica) ReplicaID() roachpb.ReplicaID {
 
 // cleanupFailedProposal cleans up after a proposal that has failed. It
 // clears any references to the proposal and releases associated quota.
-// It requires that both Replica.mu and Replica.raftMu are exclusively held.
+// It requires that Replica.mu is exclusively held.
 func (r *Replica) cleanupFailedProposalLocked(p *ProposalData) {
-	r.raftMu.AssertHeld()
 	r.mu.AssertHeld()
 	delete(r.mu.proposals, p.idKey)
 	p.releaseQuota()

--- a/pkg/kv/kvserver/store_raft.go
+++ b/pkg/kv/kvserver/store_raft.go
@@ -599,8 +599,8 @@ func (s *Store) processRaft(ctx context.Context) {
 	// spans in them, and we don't want to be leaking any.
 	s.stopper.AddCloser(stop.CloserFn(func() {
 		s.VisitReplicas(func(r *Replica) (more bool) {
-			r.mu.proposalBuf.FlushLockedWithoutProposing(ctx)
 			r.mu.Lock()
+			r.mu.proposalBuf.FlushLockedWithoutProposing(ctx)
 			for k, prop := range r.mu.proposals {
 				delete(r.mu.proposals, k)
 				prop.finishApplication(

